python/pip_install/pip_repository.bzl

```diff
@@ -278,8 +280,38 @@ def _pip_repository_impl(rctx):
 
     bzl_packages = dict(sorted([[name, normalize_name(name)] for name, _ in parsed_requirements_txt.requirements]))
 
+    # Normalize cycles first
+    requirement_cycles = {
+        name: sorted(sets.to_list(sets.make(deps)))
+        for name, deps in rctx.attr.requirement_cycles.items()
+    }
+
+    # Check for conflicts between cycles _before_ we normalize package names so
+    # that reported errors use the names the user specified
+    for i in range(len(requirement_cycles)):
+        left_group = requirement_cycles.keys()[i]
+        left_deps = requirement_cycles.values()[i]
+        for j in range(len(requirement_cycles) - (i + 1)):
+            right_deps = requirement_cycles.values()[1 + i + j]
+            right_group = requirement_cycles.keys()[1 + i + j]
+            for d in left_deps:
+                if d in right_deps:
+                    fail("Error: Requirement %s is repeated between cycles %s and %s, they must be merged." % (d, left_group, right_group))
```

Comment:
No that I can think of because the repeated dependency becomes ambiguous. If a user specifies `@pypi//foo:pkg` which is a member of two separate cycles, how do we determine which cycle the user intended to pull in? I think that only makes sense in a world where we have a way to do explicit cycle breaking (via patching or a separate flag) rather than cycle consolidation consolidation which is the core of this approach.

Reply:
Mostly thinking out loud here but the only reasonable thing we could do would be to transparently merge cycles with shared requirements on behalf of the user. It'd be nice to be a bit more declarative but Bazel generally errs on the side of explicit so it'd be out of place.